### PR TITLE
UCP/RNDV: Add get-pipeline protocol

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Implement 2-stage pipeline over get_zcopy with new protocols

### CODE REVIEW

**Comment by yosefe** [comment]
@dmitrygx @brminich can you pls take a look?

File: src/ucp/rndv/rndv_ppln.c [hunk 1]
```diff
@@ -204,14 +205,11 @@ static ucs_status_t ucp_proto_rndv_ppln_progress(uct_pending_req_t *uct_req)
...
-        if (ucp_datatype_iter_is_end_position(&req->send.state.dt_iter,
-                                              &next_iter)) {
-            return UCS_OK;
-        }
-
         ucp_datatype_iter_copy_position(&req->send.state.dt_iter, &next_iter,
                                         UCS_BIT(UCP_DATATYPE_CONTIG));
     }
+
+    return UCS_OK;
```
Comment (brminich) [inline src/ucp/rndv/rndv_ppln.c:212]: btw, is there a case when need to return not ok? Maybe make it void?

File: src/ucp/rndv/rndv_mtype.inl [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                  ucs_memory_type_names[mem_type], mtype_ep, lane);
+
+    ucp_proto_completion_init(&req->send.state.uct_comp, comp_func);
+
+    /* Set up IOV pointing to the mdesc */
+    iov.buffer     = mdesc + 1;
+    iov.length     = req->send.state.dt_iter.length;
+    iov.memh       = ucp_memh2uct(mdesc->memh, md_index);
+    iov.count      = 1;
+    iov.stride     = 0;
```
Comment (brminich) [inline src/ucp/rndv/rndv_mtype.inl:111]: extra spaces before `=`

File: src/ucp/rndv/rndv_get.c [hunk 3]
```diff
@@ -65,35 +104,39 @@ ucp_proto_rndv_get_zcopy_fetch_completion(uct_completion_t *uct_comp)
...
-        ucp_datatype_iter_t *next_iter)
+static ucs_status_t
+ucp_proto_rndv_get_zcopy_init(const ucp_proto_init_params_t *init_params)
 {
-    ucp_rkey_h rkey    = req->send.rndv.rkey;
-    uct_rkey_t tl_rkey = rkey->tl_rkey[lpriv->super.rkey_index].rkey.rkey;
-    uct_iov_t iov;
+    if (!ucp_proto_rndv_op_check(init_params, UCP_OP_ID_RNDV_RECV, 0)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
```
Comment (brminich) [inline src/ucp/rndv/rndv_get.c:115]: maybe move it to `ucp_proto_rndv_get_common_init`?

File: src/ucp/rndv/rndv_mtype.inl [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG);
+
+    iov->length = ucp_datatype_iter_next(&req->send.state.dt_iter, max_payload,
+                                         next_iter);
+    iov->buffer = UCS_PTR_BYTE_OFFSET(mdesc + 1,
+                                      req->send.state.dt_iter.offset);
+    iov->memh   = ucp_proto_rndv_mtype_get_memh(mdesc, lpriv->super.memh_index);
+    iov->stride = 0;
+    iov->count  = 1;
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_mtype.inl:80]: can be the common function with the code below?
```
void ucp_proto_rndv_mtype_iov_setup(uct_iov_t *iov, uct_mem_h memh, ucp_mem_desc_t *mdesc, size_t length, size_t offset)
{
    iov->length = length;
    iov->buffer = UCS_PTR_BYTE_OFFSET(mdesc + 1, offset);
    iov->memh   = memh;
    iov->stride = 0;
    iov->count  = 1;
}
```
and use it:
1. `ucp_proto_rndv_mtype_next_iov()`
```
ucp_proto_rndv_mtype_iov_setup(iov,
                               ucp_proto_rndv_mtype_get_memh(mdesc,
                                                             lpriv->super.memh_index),
                               mdesc,
                               ucp_datatype_iter_next(&req->send.state.dt_iter,
                                                      max_payload,
                                                      next_iter),
                               req->send.state.dt_iter.offset);
```
2. `ucp_proto_rndv_mtype_copy()`
```
ucp_proto_rndv_mtype_iov_setup(&iov, ucp_memh2uct(mdesc->memh, md_index), mdesc,
                               req->send.state.dt_iter.length, 0);
```

File: src/ucp/rndv/rndv_mtype.inl [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+{
+    ucp_worker_h worker = init_params->worker;
+    ucs_status_t status;
+
+    if ((init_params->select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        (worker->mem_type_ep[init_params->select_param->mem_type] == NULL)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (mdesc_md_map_p != NULL) {
```
Comment (dmitrygx) [inline src/ucp/rndv/rndv_mtype.inl:26]: is it used somewhere with `NULL` passed as a second parameter?
can this check be removed?

File: src/ucp/core/ucp_mm.c [hunk 6]
```diff
@@ -721,6 +721,35 @@ void ucp_frag_mpool_free(ucs_mpool_t *mp, void *chunk)
...
+    UCP_THREAD_CS_ENTER(&context->mt_lock);
+
+    if (!context->alloc_md_map_initialized) {
+        /* Allocate dummy 1-byte buffer to get the expected md_map */
+        status = ucp_mem_map_common(context, NULL, 1, UCS_MEMORY_TYPE_HOST,
+                                    UCT_MD_MEM_ACCESS_ALL, 1,
+                                    "get_alloc_md_map", &memh);
+        if (status != UCS_OK) {
+            UCP_THREAD_CS_EXIT(&context->mt_lock);
+            return status;
```
Comment (dmitrygx) [inline src/ucp/core/ucp_mm.c:739]: `got out;`

File: src/ucp/core/ucp_mm.c [hunk 7]
```diff
@@ -721,6 +721,35 @@ void ucp_frag_mpool_free(ucs_mpool_t *mp, void *chunk)
...
+
+        context->alloc_md_map_initialized = 1;
+        context->alloc_md_map             = memh->md_map;
+        ucp_mem_unmap_common(context, memh);
+    }
+
+    *md_map_p = context->alloc_md_map;
+
+    UCP_THREAD_CS_EXIT(&context->mt_lock);
+    return UCS_OK;
```
Comment (dmitrygx) [inline src/ucp/core/ucp_mm.c:750]:
```
    status - UCS_OK;

out:
    UCP_THREAD_CS_EXIT(&context->mt_lock);
    return status;
```

File: src/ucp/rndv/rndv_ppln.c (hunk 1 above)
Comment (yosefe) [reply src/ucp/rndv/rndv_ppln.c:212]: it's a UCT progress function, so it must return status

**Review by brminich — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Pull request merged**